### DISCUSSION TITLE
Chapter 08. 데이터베이스 설계

### DISCUSSION THREAD

**Discussion by crtEvent** [CodeSquad-2023-BE-Study/DB-Study#12] [완료]
책을 읽으면서 궁금한 점을 남겨주세요.

**Comment by crtEvent** [comment]
책 300p(PDF 298p)
## 규칙 2 : 다대다n：m 관계는 릴레이션으로 변환한다
> 그런 다음 두 외래키를 조합하여 주문 릴레이션의 기본키로 지정한다. 필요하다면 주문번호 속성 과 같은 별도의 기본키를 지정할 수도 있다.

- 두 외래키를 조합하여 기본키로 두는 것은 관계를 더 명확하게 표현할 수 있어서 일까요? 별도의 PK를 두는 이유?
    - 복합키보단 단일키가 select하기에도 편하고 인덱스 타기에도 좋다///
    - [복합키 단점](https://woo-chang.tistory.com/43)
- 이것 말고 또 있을까요?

**Reply by new-pow** [reply]
리뷰어한테 들을 때는 테이블 공통점을 위해서라고도 들었고,
기본키는 정의상 변경이 힘들기 때문에 외래키를 기본키 구성으로 사용하기에는 많이 불편할 것 같다는 개인적인 의견입니다.

**Reply by leegyeongwhan** [reply]
저도 궁금해서 찾아보니 논쟁이 많은 주제 같네요 복합키 vs 대체 Pk를 사용 각자의 장단점이 있는 것 같은데 직접 사용해 보지 못해
많이 와 닿지는 않네요.

**Reply by jinny-l** [reply]
새로운 신기한!

**Reply by new-pow** [reply]
https://techblog.woowahan.com/2595/

**Reply by crtEvent** [reply]
이린: pk를 따로 두는 것 관계 변경에 유리할 것 같다

**Reply by new-pow** [reply]
## 복합키 사용 예시
- 도서관이란 테이블
- 지역 + 도서관이름  -> 책

---

- 이렇게 하면 지역 검색, 지역 + 도서관 검색으로 할 때 인덱스를 탈 수 있지 않을까?
- 만약 통계적으로 쓰고 싶으면 다른 join을 하지 않고 조회하고 싶음

**Reply by new-pow** [reply]
https://woo-chang.tistory.com/43

**Reply by crtEvent** [reply]
복합키 인덱스가 궁금함 ㅎ

**Reply by crtEvent** [reply]
https://woo-chang.tistory.com/43 
jinny

**Comment by crtEvent** [comment]
책 303p
> 규칙 4   : 일대일 관계를 외래키로 표현한다
>   규칙 4-1 : 일반적인 일대일 관계는 외래키를 서로 주고 받는다
>   규칙 4-2 : 일대일 관계에 필수적으로 참여하는 개체의 릴레이션만 외래키를 받는다
>   규칙 4-3 : 모든 개체가 일대일 관계에 필수적으로 참여하면 릴레이션 하나로 합친다


- 일반적인 1:1 관계일 때 n:m 처럼 혼인 릴레이션을 따로 만들고, 남자_id, 여자_id를 외래키로 두어서 UNIQUE 조건 걸면 별로인가요?
- 뒤에 나오네요 

> '릴레이션의 개수가 많아져 릴레이션을 관리해야 하는 DBMS의 부담이 커지게 된다.외래키만으로도 표현이 가능한 1:1, 1:n 관계는 릴릴레이션으로 변환하지 않는 것이 좋다.'

- 여기서 말하는 '릴레이션을 관리해야 하는 DBMS의 부담'은 사용자가 관리하기 힘들다는 뜻일까요? 실제로 DBMS가 힘들어 한다는 뜻일까요?

**Reply by new-pow** [reply]
- 실제로 DBMS가 관리하기 힘들다는 것  같아요.
   - relation 이 늘어나면 DBMS가 저장하고 있는 자원이 많아지니까? 라고 직관적으로 생각했는데 GPT가 다음과 같이 더 정리해서 설명해주네요.

---

DBMS (Database Management System)에서 관계 (Relation)는 데이터를 저장하고 조작하는 데 사용되는 핵심 개념 중 하나입니다. 관계형 데이터베이스에서 데이터는 테이블 형식으로 구성되며, 각 테이블은 레코드(행)와 속성(열)으로 구성됩니다. 이러한 테이블은 관계로서 상호 연관된 데이터의 집합을 나타내며, 이를 효과적으로 관리하기 위해 DBMS가 사용됩니다.

DBMS에서 관계(테이블)를 관리하는 기본적인 개념과 작업은 다음과 같습니다:

1. **테이블(릴레이션) 생성**: 데이터를 저장하기 위해 테이블을 생성합니다. 각 테이블은 고유한 이름과 열(속성)을 가지며, 각 열은 해당하는 데이터 유형과 제약 조건을 가질 수 있습니다.

2. **데이터 삽입**: 테이블에 데이터를 삽입하여 레코드를 생성합니다. 각 레코드는 열에 해당하는 값을 가지며, 이러한 값들은 데이터 유형과 제약 조건에 따라 유효해야 합니다.

3. **데이터 검색**: SQL(Structured Query Language) 쿼리를 사용하여 테이블에서 데이터를 검색합니다. SELECT 문을 사용하여 원하는 조건에 따라 데이터를 필터링하고 추출할 수 있습니다.

4. **데이터 업데이트**: 데이터를 업데이트하려면 UPDATE 문을 사용하여 테이블의 레코드를 수정할 수 있습니다.

5. **데이터 삭제**: 특정 조건에 따라 테이블의 레코드를 삭제하기 위해 DELETE 문을 사용할 수 있습니다.

6. **데이터 조작(Manipulation)**: 데이터를 조작하고 변환하기 위해 다양한 SQL 함수와 연산자를 사용할 수 있습니다. 예를 들어, 데이터의 합계, 평균 등을 계산하거나 데이터를 그룹화할 수 있습니다.

7. **데이터 무결성 유지**: DBMS는 데이터 무결성을 유지하기 위해 제약 조건(Primary Key, Foreign Key, Unique, Not Null 등)을 사용합니다. 이를 통해 데이터의 정확성과 일관성을 보장합니다.

8. **인덱싱**: 테이블의 데이터에 대한 검색 성능을 향상시키기 위해 인덱스를 생성할 수 있습니다. 인덱스는 특정 열(또는 열의 조합)에 대한 빠른 검색을 가능하게 합니다.

9. **트랜잭션 관리**: 여러 데이터 조작 작업을 묶어서 하나의 논리적 작업 단위로 관리할 수 있습니다. 트랜잭션은 데이터의 일관성과 무결성을 보장하기 위해 중요합니다.

10. **보안 및 접근 제어**: DBMS는 데이터에 대한 접근 권한과 보안 설정을 관리하여 불법적인 접근과 데이터 유출을 방지합니다.

DBMS는 이러한 관계 관리 작업을 효율적으로 수행하고 데이터의 안정성과 일관성을 유지하는 역할을 합니다. 관계형 데이터베이스 시스템의 기본적인 원리와 개념을 이해하면 데이터의 구조, 조작, 관리 등을 보다 효과적으로 수행할 수 있습니다.

**Reply by new-pow** [reply]
별개의 이야긴데, 어제 새로운 신기한 예시를 발견했어요! hibernate user guide 에서
- 이렇게 hibernate매핑하면, 아래와 같이 쿼리를 내보내나봐요 (DDL auto create일때)

```
@Entity(name = "Person")
public static class Person {

	@Id
	@GeneratedValue
	private Long id;

	@OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
	private List<Phone> phones = new ArrayList<>();

	//Getters and setters are omitted for brevity

}

@Entity(name = "Phone")
public static class Phone {

	@Id
	@GeneratedValue
	private Long id;

	@Column(name = "`number`")
	private String number;

	//Getters and setters are omitted for brevity

}
```

```
CREATE TABLE Person (
    id BIGINT NOT NULL ,
    PRIMARY KEY ( id )
)

CREATE TABLE Person_Phone (
    Person_id BIGINT NOT NULL ,
    phones_id BIGINT NOT NULL
)

CREATE TABLE Phone (
    id BIGINT NOT NULL ,
    number VARCHAR(255) ,
    PRIMARY KEY ( id )
)
```

**Reply by jinny-l** [reply]
hibernate 아직 안배웠어 잉

**Comment by crtEvent** [comment]
책 318p
- 물리적 설계 하는 것도 보여줘!!

**Reply by jinny-l** [reply]
ㅋㅋㅋㅋㅋㅋ

**Reply by new-pow** [reply]
ㅋㅋㅋㅋㅋㅋㅋ GPT

**Comment by leegyeongwhan** [comment]
289P
관계는 개체 간의 의미 있는
연관성 이다. 일반적으로 관계는 요구 사항을 표현한 문장에서 동사로 표현된다.

개체의 속성을 요구사항을 분석해 추출한 결과를 토대로 개체 끼리의 관계를 동사를 통해 추출하는 과정이  저희가 지난 오브젝트 스터디를 하며 객체의 책임 역할을 공부할 때 중요한 것은 행동이다 라는 말이 생각이 나네요.  개체들의 관계에서 각 개체들이 서로 어떠한 행동을 하는지 (저는 어떠한  행동이 책에서 말하는 동사라고 생각했습니다)  분석을 하고 여기서 개체들간의 1:N ,N:M 등의 관계가 나온다고 이해했습니다.  결국 요구사항 분석에서 나오는 개념적 모델링이 중요한 것은 이 모델링을 토대로 좋은 논리적 모델링이나, 물리적 모델링이 나오고 이것을 토대로  우리가 실제 도메인 개발을 할 때 각 도메인들이 어떤 관계를 맺고 있고 , 어떤 역할을 하는지 즉 좋은 도메인이 나오려면 그전 단계인 개념적 모델링에서 부터의 각 개체들의 관계나 , 동사를 잘 모델링해야 좋은 도메인이 나온다고 이해했습니다.

**Reply by new-pow** [reply]
동사... verb...

**Reply by jinny-l** [reply]
어 이거 ajax인가 댓글 다면 바로 업데이트되네요

**Reply by new-pow** [reply]
저번주에 에이프랑 이걸로 한참 신기해햇어요

**Reply by new-pow** [reply]
ping 날라가요

<img width="486" alt="image" src="https://github.com/CodeSquad-2023-BE-Study/DB-Study/assets/103120173/bbe26ff4-5ed7-4906-9e1e-808a0d889896">

**Reply by jinny-l** [reply]
pong

**Reply by jinny-l** [reply]
깃허브 새로 고침 안해도 바로 업데이트 되는건 이거 밖에 못 본 것 같아요
